File: Third_Year/Semester1/Mobile/Lab2(native)/app/src/main/java/com/example/expensewizard/data/TransactionViewModel.kt
package com.example.expensewizard.data

import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import com.example.expensewizard.model.Transaction
import com.example.expensewizard.model.Type
import java.util.Date

class TransactionViewModel : ViewModel() {
    var transactions = mutableStateOf(listOf<Transaction>())

    init {
        populateList()
    }

    fun addTransaction(transaction: Transaction) {
        val updatedList = transactions.value + transaction
        transactions.value = updatedList
    }

    fun deleteTransaction(transactionId: Int) {
        val updatedList = transactions.value.filterNot { it.id == transactionId }
        transactions.value = updatedList
    }

    fun updateTransaction(updatedTransaction: Transaction) {
        val updatedList = transactions.value.map { transaction ->
            if (transaction.id == updatedTransaction.id) {
                updatedTransaction
            }
            else {
                transaction
            }
         }

        transactions.value = updatedList
    }

    fun getTransactionById(id : Int) : Transaction? {
        for (t in transactions.value) {
            if (t.id == id) {
                return t
            }
        }
        return Transaction("",Type.INCOME, "",0.0,Date())
    }

    private fun populateList() {
        val initialTransactions = listOf<Transaction>(
            Transaction("Uber", Type.EXPENSE,"Transport",8.76, Date()),
            Transaction("Salary", Type.INCOME,"Income",500.0, Date(2023,7,4)),
            Transaction("Groceries", Type.EXPENSE,"Food",25.46, Date(2023,8,1)),
            Transaction("Some moneyyy", Type.INCOME,"Income",50.0, Date(2023,9,29)),
            Transaction("Date", Type.EXPENSE,"Entertainment",65.65, Date(2023,10,14)),
            Transaction("Bani de la tati", Type.INCOME,"Income",400.0, Date(2023,10,15)),
            Transaction("Train", Type.EXPENSE,"Transport",2.5, Date(2023,7,20)),

        )
        transactions.value = initialTransactions

    }
}

File: Third_Year/Semester1/Mobile/Lab3(non-native)/non_native/lib/data/transaction_view_model.dart

import 'package:flutter/material.dart';

import '../model/transaction.dart';
import '../model/type.dart';
import '../ui/components/list_item.dart';
import '../ui/pages/add_page.dart';
import '../ui/pages/update_page.dart';

class TransactionViewModel extends StatefulWidget {
  const TransactionViewModel();

  @override
  _TransactionViewModelState createState() => _TransactionViewModelState();

  Function getAdd() {
    return _TransactionViewModelState().addTransaction;
  }
}

class _TransactionViewModelState extends State<TransactionViewModel> {
  List<Transaction> transactions = [];

  @override
  void initState() {
    super.initState();
    populateList();
  }

  void addTransaction(Transaction transaction) {
    setState(() {
      transactions.add(transaction);
    });

  }

  void deleteTransaction(int transactionId) {
    setState(() {
      transactions.removeWhere((transaction) => transactionId == transaction.id);
    });
  }

  void updateTransaction(Transaction updatedTransaction) {
    setState(() {
      final index = transactions.indexWhere((transaction) => transaction.id == updatedTransaction.id);
      if (index != -1) {
        transactions[index] = updatedTransaction;
      }
    });
  }

  Transaction getTransactionById(int id) {
    return transactions.firstWhere((transaction) => transaction.id == id, orElse: () {
      return Transaction(id:0,title: "",type: Type.income, category: "",amount: 0.0, timestamp:DateTime.now());
    });
  }

  void populateList() {
    transactions.addAll(List<Transaction>.of(
      [Transaction.createTransaction( "Uber", Type.expense,"Transport", 8.76,  DateTime.now()),
        Transaction.createTransaction("Salary", Type.income, "Income", 500.0, DateTime(2023, 7, 4)),
        Transaction.createTransaction("Groceries",  Type.expense,  "Food",  25.46,  DateTime(2023, 8, 1)),
        Transaction.createTransaction( "Some moneyyy",Type.income, "Income",  50.0,  DateTime(2023, 9, 29)),
        Transaction.createTransaction("Date",  Type.expense,  "Entertainment",  65.65,  DateTime(2023, 10, 14)),
        Transaction.createTransaction("Bani de la tati",Type.income,"Income", 400.0,DateTime(2023, 10, 15)),
        Transaction.createTransaction( "Train", Type.expense, "Transport", 2.5, DateTime(2023, 7, 20))
      ],

    ));
  }

  @override
  Widget build(BuildContext context) {
    return Expanded(
      child: ListView.builder(
        padding: const EdgeInsets.all(16.0),
        itemCount: transactions.length,
        itemBuilder: (context, index) {
          final transaction = transactions[index];

          if (index == 0) {
            return Column(
              children: [
                ElevatedButton(
                  onPressed: () {
                    Navigator.push(context,
                        MaterialPageRoute(builder: (_) => AddPage(addFunction: addTransaction)));
                  },
                  style: ElevatedButton.styleFrom(
                    backgroundColor: const Color(0xFFB89EFB),
                    padding: const EdgeInsets.all(20.0),
                    minimumSize: const Size(200.0, 60.0),
                  ),
                  child: const Text(
                    "+ Another One",
                    style: TextStyle(
                      color: Colors.black,
                      fontSize: 18.0,
                    ),
                  ),
                ),
                const SizedBox(height: 20.0),
                TransactionItem(
                  transaction: transaction,
                  onDelete: () => deleteTransaction(transaction.id),
                  onEdit: () {
                    final transactionId = transaction.id;
                    Navigator.push(
                      context,
                      MaterialPageRoute(builder: (_) => UpdatePage(onEdit: updateTransaction,transaction: transaction)),
                    );
                  },
                ),


              ],
            );
          }

          return Column(
            children: [
              TransactionItem(
                transaction: transaction,
                onDelete: () => deleteTransaction(transaction.id),
                onEdit: () {
                  final transactionId = transaction.id;
                  Navigator.push(
                    context,
                    MaterialPageRoute(builder: (_) => UpdatePage(onEdit: updateTransaction,transaction: transaction)),
                  );
                },
              ),
              const SizedBox(height: 16.0),
            ],
          );
        },
      ),
    );
  }
}

